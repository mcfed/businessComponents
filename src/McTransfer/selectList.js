import React from 'react';
import PureRenderMixin from 'rc-util/lib/PureRenderMixin';
import {List} from 'react-virtualized';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import {Checkbox} from 'antd';
import _ from 'lodash';

import Item from './item';
import Search from './search';
import prefixCls from './constants';

export function noop() {}

export function isRenderResultPlainObject(result) {
  return (
    result &&
    !React.isValidElement(result) &&
    Object.prototype.toString.call(result) === '[object Object]'
  );
}

export default class SelectList extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      filter: '',
      dataSource: []
    };

    this.renderItem = this.renderItem.bind(this);
    this.rowRenderer = this.rowRenderer.bind(this);
    this.handleSelect = this.handleSelect.bind(this);
    this.handleSelectAll = this.handleSelectAll.bind(this);
    this.handleFilterWapper = this.handleFilterWapper.bind(this);
    this.handleFilterWithDebounce = _.debounce(
      this.handleFilter.bind(this),
      200
    );
    this.handleFilter = this.handleFilter.bind(this);
    this.handleClear = this.handleClear.bind(this);
    this.matchFilter = this.matchFilter.bind(this);
  }

  componentWillMount() {
    this.setState({
      dataSource: this.props.dataSource
    });
  }

  componentWillReceiveProps(nextProps) {
    /* istanbul ignore else */
    if (nextProps.dataSource !== this.props.dataSource) {
      if (this.state.filter !== '') {
        this.handleFilter(nextProps.dataSource, this.state.filter);
      } else {
        this.setState({
          dataSource: nextProps.dataSource
        });
      }
    }
  }

  shouldComponentUpdate(nextProps, nextState) {
    /* istanbul ignore else */
    if (
      PureRenderMixin.shouldComponentUpdate.apply(this, nextProps, nextState)
    ) {
      return true;
    }
    return false;
  }

  componentDidUpdate() {
    /**
     * Sometimes there will be called redundant, but call forceUpdateGrid there every time is
     * the easyest way to ensure right update
     */
    this.list.forceUpdateGrid();
  }

  getCheckStatus() {
    const {selectedKeys} = this.props;
    const {dataSource} = this.state;
    /* istanbul ignore else */
    if (selectedKeys.length === 0) {
      return 'none';
    } else if (
      dataSource.every(
        item => item.disabled || selectedKeys.indexOf(item.key) >= 0
      )
    ) {
      return 'all';
    }
    return 'part';
  }

  handleSelect(selectedItem) {
    const {selectedKeys} = this.props;
    const hoder = [...selectedKeys];
    const index = hoder.indexOf(selectedItem.key);
    if (index > -1) {
      hoder.splice(index, 1);
    } else {
      hoder.push(selectedItem.key);
    }
    this.props.handleSelect(hoder);
  }

  handleSelectAll(checkAll) {
    const {dataSource} = this.state;
    const {selectedKeys} = this.props;
    const hoder = [...selectedKeys];
    let index;
    if (!checkAll) {
      dataSource.map(item => {
        /* istanbul ignore else */
        if (!item.disabled && hoder.indexOf(item.key) < 0) {
          hoder.push(item.key);
        }
        return item;
      });
    } else {
      dataSource.map(item => {
        index = hoder.indexOf(item.key);
        /* istanbul ignore else */
        if (index > -1) {
          hoder.splice(index, 1);
        }
        return item;
      });
    }
    this.props.handleSelect(hoder);
  }

  handleFilterWapper(e) {
    this.handleFilterWithDebounce(this.props.dataSource, e);
    this.setState({
      filter: e
    });
  }

  matchFilter(filter, item) {
    /* istanbul ignore else */
    if (this.props.filterOption) {
      return this.props.filterOption(filter, item);
    }
    const {renderedText} = this.renderItem(item);
    return renderedText.indexOf(filter) >= 0;
  }

  handleFilter(dataSource, filter) {
    const showItems = [];
    dataSource.map(item => {
      /* istanbul ignore else */
      if (!this.matchFilter(filter, item)) {
        return null;
      }
      showItems.push(item);
      return item;
    });
    this.setState(
      {
        dataSource: showItems
      },
      () => {
        /* TODO: maybe we can scroll to the position which user is looking at*/
        this.list.scrollToRow(0);
      }
    );
  }

  handleClear() {
    this.setState({
      filter: '',
      dataSource: this.props.dataSource
    });
  }

  rowRenderer({_key, index, _isScrolling, _isVisible, _parent, style}) {
    const {selectedKeys, header, type, rowKey, mode} = this.props;
    const item = this.state.dataSource[index];
    const {renderedText, renderedEl} = this.renderItem(item);
    const checked = selectedKeys.indexOf(item.key) >= 0;
    const itemPrefixCls = `${prefixCls}-list`;

    /* istanbul ignore else */
    if (rowKey) {
      item.key = rowKey(item);
    }

    return (
      <Item
        key={item.key}
        item={item}
        checked={checked}
        style={style}
        renderedText={renderedText}
        renderedEl={renderedEl}
        disabled={item.disabled}
        onClick={this.handleSelect}
        prefixCls={itemPrefixCls}
        header={header}
        type={type}
        mode={mode}
      />
    );
  }

  renderItem(item) {
    /* istanbul ignore next */
    const {render = noop} = this.props;
    const renderResult = render(item);
    const isRenderResultPlain = isRenderResultPlainObject(renderResult);
    return {
      renderedText: isRenderResultPlain ? renderResult.value : renderResult,
      renderedEl: isRenderResultPlain ? renderResult.label : renderResult
    };
  }

  /* istanbul ignore next */
  render() {
    const {
      footer,
      showSearch,
      showHeader,
      selectedKeys,
      itemUnit,
      itemsUnit,
      titleText,
      style,
      notFoundContent,
      searchPlaceholder,
      header,
      type,
      mode
    } = this.props;
    const {dataSource} = this.state;

    const className = classNames({
      [`${prefixCls}-list`]: true
    });

    const footerDom = footer(Object.assign({}, this.props));

    const listFooter = footerDom ? (
      <div className={`${prefixCls}-list-footer`}>{footerDom}</div>
    ) : null;

    const checkStatus = this.getCheckStatus();
    const checkedAll = checkStatus === 'all';

    const checkAllCheckbox = (
      <Checkbox
        checked={checkedAll}
        indeterminate={checkStatus === 'part'}
        onChange={() => this.handleSelectAll(checkedAll)}
      />
    );
    /* istanbul ignore next */
    const unit = dataSource.length > 1 ? itemsUnit : itemUnit;

    // height is not 100%, so there should minus 2px of the boder of transfer-list
    let bodyHeight = style.height - 2;
    if (type === 'right' && mode === 'table') {
      bodyHeight = bodyHeight - 32;
    }
    bodyHeight = showHeader ? bodyHeight - 33 : bodyHeight;
    bodyHeight = showSearch ? bodyHeight - 38 : bodyHeight;
    bodyHeight = listFooter !== null ? bodyHeight - 32 : bodyHeight;

    var listHeader =
      header &&
      header.map((value, i) => {
        return (
          <div key={`${value.dataIndex}${i}`}>
            {value.text || value.dataIndex}
          </div>
        );
      });

    const transHeader = showHeader ? (
      <div className={`${prefixCls}-list-header`}>
        {checkAllCheckbox}
        {/* {type === "left" || mode !== "table" ? ( */}
        <span className={`${prefixCls}-list-header-selected`}>
          <span>
            {(selectedKeys.length > 0 ? `${selectedKeys.length}/` : '') +
              dataSource.length}{' '}
            {unit}
          </span>
          <span className={`${prefixCls}-list-header-title`}>{titleText}</span>
        </span>
        {/* ) : (
          <div className="header-item">{listHeader}</div>
        )} */}
      </div>
    ) : null;

    const tableHeader =
      type === 'right' && mode === 'table' ? (
        <div className='header-item'>{listHeader}</div>
      ) : null;

    const search = showSearch ? (
      <Search
        searchRender={this.props.searchRender}
        value={this.state.filter}
        onChange={this.handleFilterWapper}
        handleClear={this.handleClear}
        prefixCls={`${prefixCls}-list-search`}
        placeholder={searchPlaceholder}
      />
    ) : null;

    return (
      <div className={className} style={style}>
        {transHeader}
        {search}
        {tableHeader}
        <List
          ref={list => {
            /* istanbul ignore next */
            this.list = list;
          }}
          height={dataSource.length === 0 ? 0 : bodyHeight}
          rowCount={dataSource.length}
          rowHeight={this.props.rowHeight}
          rowRenderer={this.rowRenderer}
          width={1}
          className={`${prefixCls}-list-virtualized`}
        />
        {dataSource.length === 0 && (
          <div
            className={`${prefixCls}-list-body-not-found`}
            style={{
              height: `${bodyHeight}px`,
              lineHeight: `${bodyHeight}px`
            }}>
            {notFoundContent}
          </div>
        )}
        {listFooter}
      </div>
    );
  }
}

SelectList.defaultProps = {
  filterOption: undefined,
  footer: noop,
  showSearch: false,
  showHeader: true,
  itemUnit: '',
  itemsUnit: '',
  titleText: '',
  style: {
    width: 200,
    height: 300
  },
  notFoundContent: 'Not Found',
  searchPlaceholder: 'Search here',
  rowKey: undefined,
  rowHeight: 32
};

SelectList.propTypes = {
  render: PropTypes.func.isRequired,
  dataSource: PropTypes.array.isRequired,
  selectedKeys: PropTypes.array.isRequired,
  handleSelect: PropTypes.func.isRequired,
  filterOption: PropTypes.func,
  footer: PropTypes.func,
  showSearch: PropTypes.bool,
  showHeader: PropTypes.bool,
  itemUnit: PropTypes.string,
  itemsUnit: PropTypes.string,
  titleText: PropTypes.string,
  rowHeight: PropTypes.number,
  style: PropTypes.shape({
    height: PropTypes.number.isRequired, // not support %
    width: PropTypes.any
  }),
  notFoundContent: PropTypes.string,
  searchPlaceholder: PropTypes.string,
  rowKey: PropTypes.func
};
