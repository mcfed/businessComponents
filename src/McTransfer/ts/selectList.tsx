import React from 'react';
//@ts-ignore
import PureRenderMixin from 'rc-util/lib/PureRenderMixin';
//@ts-ignore
import {List} from 'react-virtualized';
//@ts-ignore
import classNames from 'classnames';
import {Checkbox} from 'antd';
//@ts-ignore
import _ from 'lodash';

import Item from './item';
import Search from './search';
import prefixCls from './constants';
import {ItemType, ModeType} from './item';
import {ListRowProps} from 'react-virtualized/dist/es/List';
import {ItemProps, HeaderProps, StyleProps} from './commonProps';

export function noop() {}

export function isRenderResultPlainObject<T>(result: T) {
  return (
    result &&
    !React.isValidElement(result) &&
    Object.prototype.toString.call(result) === '[object Object]'
  );
}

/**
 * render：自定义渲染方法，必须
 * searchRender：自定义搜索框渲染方法，必须
 * dataSource: 数据，必须
 * selectedKeys：选中项，非必须
 * handleSelect：select方法，非必须
 * filterOption：过滤方法，非必须
 * footer：footer渲染方法，非必须
 * showSearch：是否显示搜索栏，非必须
 * showHeader: 是否显示头部，非必须
 * itemUnit: 单数单位，非必须
 * itemsUnit: 复数单位，非必须
 * titleText：title，非必须
 * rowHeight：行高，非必须
 * style: 自定义样式，非必须
 * notFoundContent: 数据为空时显示文字，非必须s
 * searchPlaceholder: 搜索栏input提示语，非必须
 * rowKey：获取key方法，必须
 * header: table模式下的column数据，table模式下必须
 * type: 区分左右模块的类型，内部使用
 * mode: 区分普通transfer和table模式，非必须，默认normal
 */

interface SelectListProps<T> {
  render: (item: object) => void;
  searchRender: () => void;
  dataSource: T[];
  selectedKeys: string[];
  handleSelect: (selectedKeys: string[]) => void;
  filterOption: (filter: string, item: object) => void;
  footer: ({}) => React.ReactNode;
  showSearch: boolean;
  showHeader: boolean;
  itemUnit: string;
  itemsUnit: string;
  titleText: string;
  rowHeight: number;
  style: StyleProps;
  notFoundContent?: string;
  searchPlaceholder?: string;
  rowKey: (item: object) => string;
  header: HeaderProps[];
  type: ItemType;
  mode: ModeType;
}

interface State<T extends ItemProps> {
  filter: string;
  dataSource: T[];
}

export default class SelectList<T extends ItemProps> extends React.Component<
  SelectListProps<T>,
  State<T>
> {
  list: null | {} = {};
  static defaultProps = {
    filterOption: undefined,
    footer: noop,
    showSearch: false,
    showHeader: true,
    itemUnit: '',
    itemsUnit: '',
    titleText: '',
    style: {
      width: 200,
      height: 300
    },
    notFoundContent: 'Not Found',
    searchPlaceholder: 'Search here',
    rowKey: undefined,
    rowHeight: 32
  };
  static state = {
    filter: '',
    dataSource: []
  };
  componentWillMount() {
    this.setState({
      dataSource: this.props.dataSource
    });
    this.shouldComponentUpdate = PureRenderMixin.shouldComponentUpdate.bind(
      this
    );
  }
  componentWillReceiveProps(nextProps: {dataSource: T[]}) {
    /* istanbul ignore else */
    if (nextProps.dataSource !== this.props.dataSource) {
      if (this.state.filter !== '') {
        this.handleFilter(nextProps.dataSource, this.state.filter);
      } else {
        this.setState({
          dataSource: nextProps.dataSource
        });
      }
    }
  }
  shouldComponentUpdate(nextProps: any, nextState: any) {
    /* istanbul ignore else */
    if (
      this.shouldComponentUpdate(nextProps, nextState)
      // PureRenderMixin.shouldComponentUpdate.apply(this, nextProps, nextState)
    ) {
      return true;
    }
    return false;
  }

  componentDidUpdate() {
    /**
     * Sometimes there will be called redundant, but call forceUpdateGrid there every time is
     * the easyest way to ensure right update
     */
    //@ts-ignore
    this.list['forceUpdateGrid']();
  }

  getCheckStatus() {
    const {selectedKeys} = this.props;
    const {dataSource} = this.state;
    /* istanbul ignore else */
    if (selectedKeys.length === 0) {
      return 'none';
    } else if (
      dataSource.every(
        item => item.disabled || selectedKeys.indexOf(item.key) >= 0
      )
    ) {
      return 'all';
    }
    return 'part';
  }

  handleSelect = (selectedItem: {key: string}) => {
    const {selectedKeys} = this.props;
    const hoder = [...selectedKeys];
    const index = hoder.indexOf(selectedItem.key);
    if (index > -1) {
      hoder.splice(index, 1);
    } else {
      hoder.push(selectedItem.key);
    }
    this.props.handleSelect(hoder);
  };

  handleSelectAll = (checkAll: boolean) => {
    const {dataSource} = this.state;
    const {selectedKeys} = this.props;
    const hoder = [...selectedKeys];
    let index;
    if (!checkAll) {
      dataSource.map((item: {key: string; disabled: boolean}) => {
        /* istanbul ignore else */
        if (!item.disabled && hoder.indexOf(item.key) < 0) {
          hoder.push(item.key);
        }
        return item;
      });
    } else {
      dataSource.map((item: {key: string}) => {
        index = hoder.indexOf(item.key);
        /* istanbul ignore else */
        if (index > -1) {
          hoder.splice(index, 1);
        }
        return item;
      });
    }
    this.props.handleSelect(hoder);
  };

  handleFilterWapper = (e: string) => {
    this.handleFilterWithDebounce(this.props.dataSource, e);
    this.setState({
      filter: e
    });
  };

  matchFilter = (filter: string, item: object) => {
    /* istanbul ignore else */
    if (this.props.filterOption) {
      return this.props.filterOption(filter, item);
    }
    const {renderedText} = this.renderItem(item);
    return renderedText.indexOf(filter) >= 0;
  };

  handleFilter = (dataSource: T[], filter: string) => {
    const showItems: T[] = [];
    dataSource.map(item => {
      /* istanbul ignore else */
      if (!this.matchFilter(filter, item)) {
        return null;
      }
      showItems.push(item);
      return item;
    });
    this.setState(
      {
        dataSource: showItems
      },
      () => {
        /* TODO: maybe we can scroll to the position which user is looking at*/
        //@ts-ignore
        this.list['scrollToRow'](0);
      }
    );
  };
  handleFilterWithDebounce = _.debounce(this.handleFilter.bind(this), 200);

  handleClear = () => {
    this.setState({
      filter: '',
      dataSource: this.props.dataSource
    });
  };

  rowRenderer(record: ListRowProps) {
    const {index, style} = record;
    const {selectedKeys, header, type, rowKey, mode} = this.props;
    const item = this.state.dataSource[index];
    const {renderedText, renderedEl} = this.renderItem(item);
    const checked = selectedKeys.indexOf(item.key) >= 0;
    const itemPrefixCls = `${prefixCls}-list`;

    /* istanbul ignore else */
    if (rowKey) {
      item.key = rowKey(item);
    }

    return (
      <Item
        key={item.key}
        item={item}
        checked={checked}
        style={style}
        renderedText={renderedText}
        renderedEl={renderedEl}
        disabled={item.disabled}
        onClick={this.handleSelect}
        prefixCls={itemPrefixCls}
        header={header}
        type={type}
        mode={mode}
      />
    );
  }

  renderItem(item: object) {
    /* istanbul ignore next */
    const {render = noop} = this.props;
    const renderResult = render(item);
    const isRenderResultPlain = isRenderResultPlainObject(renderResult);
    return {
      //@ts-ignore
      renderedText: isRenderResultPlain ? renderResult['value'] : renderResult,
      //@ts-ignore
      renderedEl: isRenderResultPlain ? renderResult['label'] : renderResult
    };
  }

  /* istanbul ignore next */
  render() {
    const {
      footer,
      showSearch,
      showHeader,
      selectedKeys,
      itemUnit,
      itemsUnit,
      titleText,
      style,
      notFoundContent,
      searchPlaceholder,
      header,
      type,
      mode
    } = this.props;
    const {dataSource} = this.state;

    const className = classNames({
      [`${prefixCls}-list`]: true
    });

    const footerDom = footer(Object.assign({}, this.props));

    const listFooter = footerDom ? (
      <div className={`${prefixCls}-list-footer`}>{footerDom}</div>
    ) : null;

    const checkStatus = this.getCheckStatus();
    const checkedAll = checkStatus === 'all';

    const checkAllCheckbox = (
      <Checkbox
        checked={checkedAll}
        indeterminate={checkStatus === 'part'}
        onChange={() => this.handleSelectAll(checkedAll)}
      />
    );
    /* istanbul ignore next */
    const unit = dataSource.length > 1 ? itemsUnit : itemUnit;

    // height is not 100%, so there should minus 2px of the boder of transfer-list
    let bodyHeight = style.height - 2;
    if (type === 'right' && mode === 'table') {
      bodyHeight = bodyHeight - 32;
    }
    bodyHeight = showHeader ? bodyHeight - 33 : bodyHeight;
    bodyHeight = showSearch ? bodyHeight - 38 : bodyHeight;
    bodyHeight = listFooter !== null ? bodyHeight - 32 : bodyHeight;

    var listHeader =
      header &&
      header.map((value, i) => {
        return (
          <div key={`${value.dataIndex}${i}`}>
            {value.text || value.dataIndex}
          </div>
        );
      });

    const transHeader = showHeader ? (
      <div className={`${prefixCls}-list-header`}>
        {checkAllCheckbox}
        {/* {type === "left" || mode !== "table" ? ( */}
        <span className={`${prefixCls}-list-header-selected`}>
          <span>
            {(selectedKeys.length > 0 ? `${selectedKeys.length}/` : '') +
              dataSource.length}{' '}
            {unit}
          </span>
          <span className={`${prefixCls}-list-header-title`}>{titleText}</span>
        </span>
        {/* ) : (
              <div className="header-item">{listHeader}</div>
            )} */}
      </div>
    ) : null;

    const tableHeader =
      type === 'right' && mode === 'table' ? (
        <div className='header-item'>{listHeader}</div>
      ) : null;

    const search = showSearch ? (
      <Search
        searchRender={this.props.searchRender}
        value={this.state.filter}
        onChange={this.handleFilterWapper}
        handleClear={this.handleClear}
        prefixCls={`${prefixCls}-list-search`}
        placeholder={searchPlaceholder}
      />
    ) : null;

    return (
      <div className={className} style={style}>
        {transHeader}
        {search}
        {tableHeader}
        <List
          ref={list => {
            /* istanbul ignore next */
            this.list = list;
          }}
          height={dataSource.length === 0 ? 0 : bodyHeight}
          rowCount={dataSource.length}
          rowHeight={this.props.rowHeight}
          rowRenderer={this.rowRenderer}
          width={1}
          className={`${prefixCls}-list-virtualized`}
        />
        {dataSource.length === 0 && (
          <div
            className={`${prefixCls}-list-body-not-found`}
            style={{
              height: `${bodyHeight}px`,
              lineHeight: `${bodyHeight}px`
            }}>
            {notFoundContent}
          </div>
        )}
        {listFooter}
      </div>
    );
  }
}
