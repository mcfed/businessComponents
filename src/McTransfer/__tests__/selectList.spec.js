import React from 'react';
import {shallow} from 'enzyme';
import SelectList from '../ts/selectList.tsx';
import * as Sl from '../ts/selectList.tsx';
import PureRenderMixin from 'rc-util/lib/PureRenderMixin';
jest.mock('rc-util/lib/PureRenderMixin');

function setup(props, options) {
  const defaultProps = {
    selectedKeys: ['a'],
    dataSource: [
      {
        key: 'a'
      }
    ],
    render: jest.fn(() => {
      return {
        test: 'string'
      };
    }),
    handleSelect: jest.fn(),
    rowHeight: 1,
    rowKey: jest.fn().mockImplementation(name => true),
    filterOption: undefined,
    showSearch: true,
    showHeader: true,
    footerDom: true,
    footer: jest.fn(() => true),
    notFoundContent: ''
  };
  const wrapper = shallow(
    <SelectList {...Object.assign({}, defaultProps, props)} />,
    options
  );

  return {
    wrapper
  };
}

describe('快照测试', () => {
  it('全页快照', () => {
    const props = {
      item: {}
    };
    const {wrapper} = setup(props, {disableLifecycleMethods: true});
    expect(wrapper).toMatchSnapshot();
  });
});

describe('方法测试全覆盖', () => {
  const props = {
    dataSource: []
  };

  it('componentWillMount 生命周期测试', () => {
    const {wrapper} = setup(props, {disableLifecycleMethods: true});
    const instance = wrapper.instance();
    instance.componentWillMount();
  });

  it('componentWillReceiveProps filter 为空 生命周期测试', () => {
    const {wrapper} = setup(props, {disableLifecycleMethods: true});
    const instance = wrapper.instance();
    instance.handleFilter = jest.fn();
    instance.state.filter = '';
    instance.componentWillReceiveProps({dataSource: [{key: 'a'}]});
  });

  it('componentWillReceiveProps filter 不为空 生命周期测试', () => {
    const {wrapper} = setup(props, {disableLifecycleMethods: true});
    const instance = wrapper.instance();
    instance.handleFilter = jest.fn();
    instance.state.filter = '1';
    instance.componentWillReceiveProps({dataSource: [{key: 'a'}]});
    const handleFilterSpy = jest.spyOn(instance, 'handleFilter');
    expect(handleFilterSpy).toHaveBeenCalled();
  });

  it('shouldComponentUpdate true branch 生命周期测试', () => {
    PureRenderMixin.shouldComponentUpdate.apply = jest.fn(() => true);
    const {wrapper} = setup(props, {disableLifecycleMethods: true});
    const instance = wrapper.instance();
    instance.shouldComponentUpdate();
  });

  it('shouldComponentUpdate false branch 生命周期测试', () => {
    PureRenderMixin.shouldComponentUpdate.apply = jest.fn(() => false);
    const {wrapper} = setup(props, {disableLifecycleMethods: true});
    const instance = wrapper.instance();
    instance.shouldComponentUpdate();
  });

  it('componentDidUpdate 生命周期测试', () => {
    const {wrapper} = setup(props, {disableLifecycleMethods: true});
    const instance = wrapper.instance();
    instance.list = {
      forceUpdateGrid: jest.fn()
    };
    instance.componentDidUpdate();
  });

  it('getCheckStatus selectedKeys 不为空 branch测试', () => {
    const {wrapper} = setup(props, {disableLifecycleMethods: true});
    const instance = wrapper.instance();
    instance.getCheckStatus();
  });

  it('getCheckStatus selectedKeys 为空 branch测试', () => {
    props.selectedKeys = [];
    const {wrapper} = setup(props, {disableLifecycleMethods: true});
    const instance = wrapper.instance();
    instance.getCheckStatus();
  });

  it('handleSelect index 大于-1 branch测试', () => {
    props.selectedKeys = ['a'];
    const {wrapper} = setup(props, {disableLifecycleMethods: true});
    const instance = wrapper.instance();
    instance.handleSelect({key: 'a'});
    expect(instance.props.handleSelect).toHaveBeenCalled();
  });

  it('handleSelect index 小于-1 branch测试', () => {
    const {wrapper} = setup(props, {disableLifecycleMethods: true});
    const instance = wrapper.instance();
    instance.handleSelect({key: 'b'});
    expect(instance.props.handleSelect).toHaveBeenCalled();
  });

  it('getCheckStatus 方法测试', () => {
    const {wrapper} = setup(props, {disableLifecycleMethods: true});
    const instance = wrapper.instance();
    instance.state.dataSource = [
      {
        key: 'b'
      }
    ];
    instance.getCheckStatus();
  });

  it('handleSelectAll true branch测试', () => {
    const {wrapper} = setup(props, {disableLifecycleMethods: true});
    const instance = wrapper.instance();
    instance.state.dataSource = [
      {
        key: 'a'
      }
    ];
    instance.handleSelectAll(true);
    expect(instance.props.handleSelect).toHaveBeenCalled();
  });

  it('handleSelectAll false branch测试', () => {
    const {wrapper} = setup(props, {disableLifecycleMethods: true});
    const instance = wrapper.instance();
    instance.state.dataSource = [
      {
        key: 'b'
      }
    ];
    instance.handleSelectAll(false);
    expect(instance.props.handleSelect).toHaveBeenCalled();
  });

  it('handleFilterWapper 方法测试', () => {
    const {wrapper} = setup(props, {disableLifecycleMethods: true});
    const instance = wrapper.instance();
    const handleFilterWithDebounceSpy = jest.spyOn(
      instance,
      'handleFilterWithDebounce'
    );
    instance.handleFilterWapper();
    expect(handleFilterWithDebounceSpy).toHaveBeenCalled();
  });

  it('matchFilter filterOption true branch测试', () => {
    props.filterOption = jest.fn(() => true);
    const {wrapper} = setup(props, {disableLifecycleMethods: true});
    const instance = wrapper.instance();
    instance.matchFilter('', '');
  });

  it('matchFilter filterOption false branch测试', () => {
    props.filterOption = undefined;
    const {wrapper} = setup(props, {disableLifecycleMethods: true});
    const instance = wrapper.instance();
    instance.renderItem = jest.fn(name => {
      return {
        renderedText: '',
        renderedEl: ''
      };
    });
    instance.matchFilter('', '');
  });

  it('handleFilter matchFilter true branch测试', () => {
    const {wrapper} = setup(props, {disableLifecycleMethods: true});
    const instance = wrapper.instance();
    instance.list = {
      scrollToRow: jest.fn()
    };
    instance.matchFilter = jest.fn().mockImplementation(() => true);
    instance.handleFilter([''], '');
  });

  it('handleFilter matchFilter false branch测试', () => {
    const {wrapper} = setup(props, {disableLifecycleMethods: true});
    const instance = wrapper.instance();
    instance.list = {
      scrollToRow: jest.fn()
    };
    instance.matchFilter = jest.fn().mockImplementation(() => false);
    instance.handleFilter([''], '');
  });

  it('handleClear 方法测试', () => {
    const {wrapper} = setup(props, {disableLifecycleMethods: true});
    const instance = wrapper.instance();
    instance.handleClear();
  });

  it('rowRenderer 方法测试', () => {
    const {wrapper} = setup(props, {disableLifecycleMethods: true});
    const instance = wrapper.instance();
    instance.state.dataSource = [
      {
        key: ''
      }
    ];
    instance.renderItem = jest.fn().mockImplementation(name => {
      return {
        renderedText: '',
        renderedEl: ''
      };
    });
    instance.rowRenderer({
      _key: '',
      index: 0,
      _isScrolling: '',
      _isVisible: '',
      _parent: '',
      style: {}
    });
  });

  it('renderItem 方法测试', () => {
    const {wrapper} = setup(props, {disableLifecycleMethods: true});
    const instance = wrapper.instance();
    instance.renderItem();
  });

  it('renderItem branch 方法测试', () => {
    props.render = jest.fn(() => true);
    const {wrapper} = setup(props, {disableLifecycleMethods: true});
    const instance = wrapper.instance();
    instance.renderItem();
  });

  it('noop 方法测试', () => {
    Sl.noop();
  });

  it('isRenderResultPlainObject 方法测试', () => {
    Sl.isRenderResultPlainObject(true);
    Sl.isRenderResultPlainObject(false);
  });

  it('showSearch 和 showHeader 等 false branch 测试', () => {
    props.showHeader = false;
    props.showSearch = false;
    props.footerDom = false;
    const {wrapper} = setup(props, {disableLifecycleMethods: true});
    const instance = wrapper.instance();
    props.showHeader = true;
    props.showSearch = true;
    props.footerDom = true;
  });

  it('Checkbox onChange方法测试', () => {
    const {wrapper} = setup(props, {disableLifecycleMethods: true});
    wrapper
      .find('Checkbox')
      .props()
      .onChange();
  });

  it('bodyHeight branch and dataSource length 大于1 测试', () => {
    props.footer = jest.fn(() => false);
    const {wrapper} = setup(props, {disableLifecycleMethods: true});
    const instance = wrapper.instance();
  });
});
