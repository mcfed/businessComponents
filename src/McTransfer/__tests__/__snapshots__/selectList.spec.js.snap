// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`快照测试 全页快照 1`] = `
<div
  className="transfer-list"
  style={
    Object {
      "height": 300,
      "width": 200,
    }
  }
>
  <div
    className="transfer-list-header"
  >
    <Checkbox
      checked={true}
      indeterminate={false}
      onChange={[Function]}
    />
    <span
      className="transfer-list-header-selected"
    >
      <span>
        1/1
         
      </span>
      <span
        className="transfer-list-header-title"
      />
    </span>
  </div>
  <Search
    handleClear={[Function]}
    onChange={[Function]}
    placeholder="Search here"
    prefixCls="transfer-list-search"
  />
  <List
    autoHeight={false}
    className="transfer-list-virtualized"
    estimatedRowSize={30}
    height={195}
    noRowsRenderer={[Function]}
    onRowsRendered={[Function]}
    onScroll={[Function]}
    overscanIndicesGetter={[Function]}
    overscanRowCount={10}
    rowCount={1}
    rowHeight={1}
    rowRenderer={[Function]}
    scrollToAlignment="auto"
    scrollToIndex={-1}
    style={Object {}}
    width={1}
  />
  <div
    className="transfer-list-footer"
  >
    <Component />
  </div>
</div>
`;
